### ISSUE TITLE
Missing lifetime specifier for the `WindowHandler` trait

### ISSUE THREAD

**Issue by d0rianb** [QuantumBadger/Speedy2D#52]
I nee to ensure that a property has the same lifetime as my `WindowHandler` implementation : 
```rust 
impl<'a> WindowHandler<EditorEvent> for EditorWindowHandler<'a>
```
But if I try to specify a lifetime I get this error : 
```
error[E0308]: method not compatible with trait
 note: expected fn pointer `fn(&mut EditorWindowHandler<'a>, &mut WindowHelper<EditorEvent>, _)`
          found fn pointer `fn(&'a mut EditorWindowHandler<'a>, &mut WindowHelper<EditorEvent>, _)`
```
I think that the problem is that the `WindowHandler` trait is declared with anonymous lifetime specifier.

**Comment by QuantumBadger** [comment]
Is it possible to make the lifetime of your property `'static`? The `WindowHandler` needs to exist until the program exits.

If that doesn't work, please could you post a more complete code sample to show what you're trying to achieve?

**Comment by d0rianb** [comment]
I tried your solution : 
```rust
struct EditorWindowHandler {
    editor: Editor<'static>,
    ...
}
```
But when it comes to to `on_keyboard_char` method of the `EditorWindowHandler` : 
```rust 
fn on_keyboard_char(&mut self, helper: &mut WindowHelper<EditorEvent>, unicode_codepoint: char) {
        self.editor.add_char(unicode_codepoint.to_string());
    }
```
I get this error message because the compiler don't know that my window handler is static : 
```
 fn on_keyboard_char(&mut self, helper: &mut WindowHelper<EditorEvent>, unicode_codepoint: char) {
    |                --------- this data with an anonymous lifetime `'_`...
159 |             self.editor.add_char(unicode_codepoint.to_string());
    |             ----------- ^^^^^^^^ ...and is required to live as long as `'static` here
````

**Comment by QuantumBadger** [comment]
Would it be possible to provide a (shrunk-down) full Rust file which demonstrates this please, which I can compile myself? Currently I'm having trouble understanding how this issue could occur, and I'm not sure that adding an explicit lifetime to the WindowHandler would fix things.

**Comment by d0rianb** [comment]
You can find my projet source code at this [url](https://github.com/d0rianb/text-editor).
Normally you will have the error message when trying to `cargo run`.

**Comment by QuantumBadger** [comment]
Thanks! I've had a look, and that code compiles correctly when the lifetime specifiers are removed.

I assume the lifetimes were added due to the lambdas in each `MenuItem` -- while a lambda-based approach should work without needing lifetime specifiers, I'd suggest going with an event-based approach instead. Each `MenuItem` could contain an "event type" enum constant, and then this could be sent to the main app through the `event_sender`.

I've attached a couple of patches that demonstrate this approach -- the first one removes the lifetime specifiers to get the code building, and the second one adds a `MenuSelection` enum and uses it for menu events. Hope this is helpful :)

[patches.zip](https://github.com/QuantumBadger/Speedy2D/files/7849828/patches.zip)
